lib/Dialect/Pipeline/PipelineOps.cpp

```diff
@@ -327,6 +327,155 @@ LogicalResult ReturnOp::verify() {
+          return failure();
+        registers.push_back(v);
+        registerTypes.push_back(t);
+        llvm::append_range(clockGates, cgs);
+        clockGatesPerRegisterList.push_back(cgs.size());
+        return success();
+      })))
+    return failure();
+
+  clockGatesPerRegister =
+      parser.getBuilder().getI64ArrayAttr(clockGatesPerRegisterList);
+
+  return success();
+}
+
+void printStageRegisters(OpAsmPrinter &p, Operation *op, ValueRange registers,
+                         TypeRange registerTypes, ValueRange clockGates,
+                         ArrayAttr clockGatesPerRegister) {
+  if (registers.empty())
+    return;
+
+  p << "regs(";
+  size_t clockGateStartIdx = 0;
+  llvm::interleaveComma(
+      llvm::zip(registers, registerTypes, clockGatesPerRegister), p,
+      [&](auto it) {
+        auto &[reg, type, nClockGatesAttr] = it;
+        p << reg << " : " << type;
+        int64_t nClockGates =
+            nClockGatesAttr.template cast<IntegerAttr>().getInt();
+        if (nClockGates == 0)
+          return;
+        p << " gated by [";
+        llvm::interleaveComma(clockGates.slice(clockGateStartIdx, nClockGates),
+                              p);
+        p << "]";
+        clockGateStartIdx += nClockGates;
+      });
+  p << ")";
+}
+
+void StageOp::build(OpBuilder &odsBuilder, OperationState &odsState,
+                    Block *dest, ValueRange registers,
+                    ValueRange passthroughs) {
+  odsState.addSuccessors(dest);
+  odsState.addOperands(registers);
+  odsState.addOperands(passthroughs);
+  odsState.addAttribute("operand_segment_sizes",
+                        odsBuilder.getDenseI32ArrayAttr(
+                            {static_cast<int32_t>(registers.size()),
+                             static_cast<int32_t>(passthroughs.size()),
+                             /*clock gates*/ static_cast<int32_t>(0)}));
+  llvm::SmallVector<int64_t> clockGatesPerRegister(registers.size(), 0);
+  odsState.addAttribute("clockGatesPerRegister",
+                        odsBuilder.getI64ArrayAttr(clockGatesPerRegister));
+}
+
+void StageOp::build(
+    OpBuilder &odsBuilder, OperationState &odsState, Block *dest,
+    ValueRange registers, ValueRange passthroughs,
+    const llvm::DenseMap<Value, llvm::SmallVector<Value>> &clockGates) {
+  odsState.addSuccessors(dest);
+  odsState.addOperands(registers);
+  odsState.addOperands(passthroughs);
+
+  // Get a sorted list of clock gates wrt. the registers.
+  llvm::SmallVector<int64_t> clockGatesPerRegister;
+  llvm::SmallVector<Value> sortedClockGates;
+  for (Value reg : registers) {
+    size_t nClockGates = 0;
+    if (auto it = clockGates.find(reg); it != clockGates.end()) {
+      llvm::append_range(sortedClockGates, it->second);
+      nClockGates = it->second.size();
+    }
+    clockGatesPerRegister.push_back(nClockGates);
+  }
+  odsState.addOperands(sortedClockGates);
+  odsState.addAttribute(
+      "operand_segment_sizes",
```

Comment:
Just for my understanding, what is `operand_segment_sizes`?

Reply:
https://github.com/llvm/llvm-project/blob/fc2c8fed0b47217814e2b917e94d2d8b185c35a2/mlir/include/mlir/IR/OpDefinition.h#L1330